sea-query-derive/src/lib.rs

```diff
@@ -121,126 +236,139 @@ where
+    }
+}
 
-    let table_name = match get_table_name(&ident, attrs) {
+#[proc_macro_attribute]
+pub fn enum_def(args: TokenStream, input: TokenStream) -> TokenStream {
+    let attr_args = match darling::ast::NestedMeta::parse_meta_list(args.into()) {
         Ok(v) => v,
-        Err(e) => return e.to_compile_error().into(),
+        Err(e) => {
+            return TokenStream::from(darling::Error::from(e).write_errors());
+        }
     };
+    let input = parse_macro_input!(input as DeriveInput);
 
-    // Currently we only support enums and unit structs
-    let variants =
-        match data {
-            syn::Data::Enum(DataEnum { variants, .. }) => variants,
-            syn::Data::Struct(DataStruct {
-                fields: Fields::Unit,
-                ..
-            }) => {
-                let impl_iden = impl_iden_for_unit_struct(&ident, &table_name);
-
-                return quote! {
-                    #impl_iden
-
-                    impl #sea_query_path::IdenStatic for #ident {
-                        fn as_str(&self) -> &'static str {
-                            #table_name
-                        }
-                    }
-
-                    impl std::convert::AsRef<str> for #ident {
-                        fn as_ref(&self) -> &str {
-                            self.as_str()
-                        }
-                    }
-                }
-                .into();
-            }
-            _ => return quote_spanned! {
-                ident.span() => compile_error!("you can only derive Iden on enums or unit structs");
-            }
-            .into(),
-        };
-
-    if variants.is_empty() {
-        return TokenStream::new();
-    }
-
-    let impl_iden = impl_iden_for_enum(&ident, &table_name, variants.iter());
+    let args = match GenEnumArgs::from_list(&attr_args) {
+        Ok(v) => v,
+        Err(e) => {
+            return TokenStream::from(e.write_errors());
+        }
+    };
 
-    let match_arms = match variants
-        .iter()
-        .map(|v| (table_name.as_str(), v))
-        .map(IdenVariant::<DeriveIdenStatic>::try_from)
-        .collect::<syn::Result<Vec<_>>>()
-    {
-        Ok(v) => quote! { #(#v),* },
-        Err(e) => return e.to_compile_error().into(),
+    let fields = match &input.data {
+        Data::Struct(DataStruct {
+            fields: Fields::Named(fields),
+            ..
+        }) => &fields.named,
+        _ => panic!("#[enum_def] can only be used on structs"),
     };
 
-    let output = quote! {
-        #impl_iden
+    let field_names: Vec<NamingHolder> = fields
+        .iter()
+        .map(|field| {
+            let ident = &field.ident;
+            let string = ident
+                .as_ref()
+                .expect("#[enum_def] can only be used on structs with named fields")
+                .to_string();
+            let as_pascal = string.to_pascal_case();
+            NamingHolder {
+                default: ident.as_ref().unwrap().clone(),
```

Comment:
That's an unnecessary unwrap, the unwrapped option should be stored before

Reply:
My final variant is the following:
```rust
    let fields =
        match &input.data {
            Data::Struct(DataStruct {
                fields: Fields::Named(fields),
                ..
            }) => &fields.named,
            _ => return quote_spanned! {
                input.span() => compile_error!("you can only derive Iden on enums or unit structs");
            }
            .into(),
        };

    let field_names: Vec<NamingHolder> = fields
        .iter()
        .map(|field| {
            let ident = field.ident.as_ref().unwrap();
            NamingHolder {
                default: ident.clone(),
                pascal: Ident::new(ident.to_string().to_pascal_case().as_str(), ident.span()),
            }
        })
        .collect();
```
I'm not sure, but Fields::Named means there are only named fields, so I've kept `unwrap` with slight changes for `ident`.